zipline/pipeline/loaders/utils.py

```diff
@@ -260,3 +274,53 @@ def check_data_query_args(data_query_time, data_query_tz):
                 data_query_tz,
             ),
         )
+
+
+def zip_with_floats(dates, flts):
+        return pd.Series(flts, index=dates, dtype='float')
+
+
+def num_days_between(dates, start_date, end_date):
+    return num_days_in_range(dates, start_date, end_date)
+
+
+def zip_with_dates(index_dates, dts):
+    return pd.Series(pd.to_datetime(dts), index=index_dates)
+
+
+def get_values_for_date_ranges(zip_date_index_with_vals,
+                               vals_for_date_intervals,
+                               date_intervals,
+                               date_index):
+    """
+    Returns a Series of values indexed by date based on values for the given
+    date intervals.
+
+    Parameters
+    ----------
+    zip_date_index_with_vals : callable
+        A function that takes in a list of dates and a list of values and
+        returns a pd.Series with the values indexed by the dates.
+    vals_for_date_intervals : list
+        A list of values for each date interval in `date_intervals`.
+    date_intervals : list
+        A list of pairs of dates, where each pair represents a date interval
+        that corresponds to the value at the same index in
+        `vals_for_date_intervals`.
+    date_index : DatetimeIndex
+        The DatetimeIndex containing all dates for which values were requested.
+
+    Returns
+    -------
+    date_index_with_vals : pd.Series
+        A Series indexed by the given DatetimeIndex and with values assigned
+        to dates based on the given date intervals.
+    """
+    # Fill in given values for given date ranges.
+    return zip_date_index_with_vals(
+        date_index,
+        np.repeat(vals_for_date_intervals,
```

Comment:
@ssanderson I've changed this a bit to make the code clearer, but I'm not sure what to do from here to get the performance benefits you mentioned.